### PR TITLE
Fixes #11926 - Authority Customizer.

### LINKED ISSUE #11926: File upload puzzle
**Jetty Version**
INFO :oejs.Server:main: jetty-12.0.10; built: 2024-05-30T04:40:36.563Z; git: https://github.com/jetty/jetty.project/commit/26106dfc84a03ddb6216062fe33b047fc332d0ce; jvm 22.0.1+8

**Jetty Environment**
core,ee8

**Java Version**
openjdk 22.0.1 2024-04-16
OpenJDK Runtime Environment Temurin-22.0.1+8 (build 22.0.1+8)
OpenJDK 64-Bit Server VM Temurin-22.0.1+8 (build 22.0.1+8, mixed mode)

**Question**
We have a very simple file upload service that worked very well with Jetty-10 but gets stuck with files larger than 12KB on Jetty-12 throwing a timeout exception.
```java
Type: IOException
Message: java.util.concurrent.TimeoutException: Idle timeout 30000 ms elapsed
```
What could be blocking file upload? The attached servlet blocks on file upload when the file size is more than 12KB i.e.  XLSX_150 works but XLSX_200 hangs. I must be missing a tunable knob.  
Please help, I am at a loss where to look. 

Thank you.
The servlet: [JettyHangingServlet.zip](https://github.com/user-attachments/files/15856075/JettyHangingServlet.zip)
Jetty Configuration: [jetty configuration](https://github.com/user-attachments/files/15856088/jetty_config.txt) 
Sample files: [file_example_XLSX_150.xlsx](https://github.com/user-attachments/files/15856078/file_example_XLSX_150.xlsx)  file_example_XLSX_200.xlsx](https://github.com/user-attachments/files/15856079/file_example_XLSX_200.xlsx)

### PR DESCRIPTION
**Pull request by sbordet**
Introduced AuthorityCustomizer to synthesize the authority from the Host header (or serverName:serverPort).

### CODE REVIEW

**Review by gregw — DISMISSED** [review]
This looks good.  But do we have the ability to specify a different HttpConfiguration for h2 vs https?   As we might want this to apply only to h2.... but then https will already set the authority from the host... so probably OK

**Comment by sbordet** [comment]
@gregw yes we have the ability to specify different `HttpConfiguration` instances to different `ConnectionFactory` instances, so we are good.

**Review by gregw — DISMISSED** [review]

File: jetty-core/jetty-http2/jetty-http2-server/src/main/java/org/eclipse/jetty/http2/server/AuthorityCustomizer.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+/**
+ * <p>A {@link HttpConfiguration.Customizer} that synthesizes the authority when the
+ * {@link HttpHeader#C_AUTHORITY} header is missing.</p>
+ * <p>After customization, the synthesized authority is accessible via
+ * {@link HttpURI#getAuthority()} from the {@link Request} object.</p>
+ * <p>The authority is synthesized from the {@code Host} header.
+ * If the {@code Host} header is also missing, it is synthesized using
+ * {@link Request#getServerName(Request)} and {@link Request#getServerPort(Request)}.</p>
+ */
+public class AuthorityCustomizer implements HttpConfiguration.Customizer
```
Comment (joakime) [inline jetty-core/jetty-http2/jetty-http2-server/src/main/java/org/eclipse/jetty/http2/server/AuthorityCustomizer.java:33]: If someone has both this, and `HostHeaderCustomizer` active at the same time, what happens?

**Review by joakime — COMMENTED** [review]
Can you add some tests where both this and `HostHeaderCustomizer` are present at the same time?

File: jetty-core/jetty-http2/jetty-http2-server/src/main/java/org/eclipse/jetty/http2/server/AuthorityCustomizer.java (hunk 1 above)
Comment (gregw) [reply jetty-core/jetty-http2/jetty-http2-server/src/main/java/org/eclipse/jetty/http2/server/AuthorityCustomizer.java:33]: Let me try answer this one to see if I understand correctly.  Two options:

 a) `HostHeaderCustomizer` then `AuthorityCustomizer` - The authority, if missing will be set from the configured host values if missing from the request.
 b) `AuthorityCustomizer` then `HostHeaderCustomizer` - Any missing host header will be constructed from the authority created by the Authority customizer (probably from the connection meta data)

**Comment by a1730** [comment]
Hi 👋 

I must be missing something on how to activate this patch.
I created an authority-customizer module  -see attached moduledef tar file, and deployed it to ```jetty-home-12.0.12-SNAPSHOT``` base.
I did not notice a change in behaviour between Jetty-12.11 and Jetty-12.12, i.e. I get a 400 missing authority response.

Please see attached module definition file and logs, and let me know what knob to turn.
Thank you.
[moduledef.tar.gz](https://github.com/user-attachments/files/16355622/moduledef.tar.gz)
[log240723.tar.gz](https://github.com/user-attachments/files/16355623/log240723.tar.gz)

File: jetty-core/jetty-http2/jetty-http2-server/src/main/java/org/eclipse/jetty/http2/server/AuthorityCustomizer.java (hunk 1 above)
Comment (sbordet) [reply jetty-core/jetty-http2/jetty-http2-server/src/main/java/org/eclipse/jetty/http2/server/AuthorityCustomizer.java:33]: @gregw is correct, the 2 customizers are orthogonal.

**Comment by sbordet** [comment]
@a1730 the problem is an additional check on the authority performed _before_ the customizers are invoked.

@gregw @joakime I think the checks on the authority may be a little too strict here:
https://github.com/jetty/jetty.project/blob/jetty-12.0.11/jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpCompliance.java#L405-407

At that point, we know the `Host` header is valid, and for HTTP/2 the authority may be missing, but that should be ok, as it is ok for the `Host` header to be missing.

File: jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpCompliance.java [hunk 2]
```diff
@@ -402,9 +402,6 @@ public static void checkHttpCompliance(MetaData.Request request, HttpCompliance
                     for (String hostValue: hostValues)
                         if (StringUtil.isBlank(hostValue))
                             assertAllowed(Violation.UNSAFE_HOST_HEADER, mode, listener);
-                    String authority = request.getHttpURI().getHost();
-                    if (StringUtil.isBlank(authority))
-                        assertAllowed(Violation.UNSAFE_HOST_HEADER, mode, listener);
```
Comment (joakime) [inline jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpCompliance.java:407]: This is just an unsafe compliance violation, id doesn't fail the header if you allow that specific violation.  Please restore these lines of code, if not here, somewhere else.

**Review by joakime — CHANGES_REQUESTED** [review]
~Need to restore HttpCompliance.Violation.UNSAFE_HOST_HEADER~

File: jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpCompliance.java (hunk 2 above)
Comment (joakime) [reply jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpCompliance.java:407]: Had conversation with @sbordet - i'm sold on the new future authority check PR

**Review by joakime — APPROVED** [review]

**Comment by sbordet** [comment]
See also #12110. This PR removed misplaced authority checks, but #12110 should restore them in the right place.

**Pull request merged**